### ISSUE TITLE
How to handle short names for Docker APIv2?

### ISSUE THREAD

**Issue by matejvasek** [containers/podman#11564]
Right now many users are troubled by default enforcing policy for short-names when using `podman` as `docker` daemon.
I had to instruct multiple users to set permissive mode when using `podman machine` on macOS.

The thing is that even if they use full name as they should `docker.io` is treated as a special case by some clients: client truncates it and query param only contains short name.

Possibilities:
* Change default setting to permissive.
* Ignore the setting in compat endpoitns and assume no-host-name==docker.io?</heresy>
* Create separate setting just for compat endpoint that would be permissive by default.

**Comment by rhatdan** [comment]
We have a patch right now, which is setting registries.conf to docker.io by default.  This is just a stop gap.  Next step is to move to prompting in the client. For podman.io.
We are also considering just having docker.io for the compatibility library.

**Comment by matejvasek** [comment]
>Next step is to move to prompting in the client.

@rhatdan I am not sure what that means. Isn't `podman` CLI already prompting to choose registry if needed? If so this is not viable for docker apiv2.

**Comment by mheon** [comment]
For Libpod APIv2, we want to take the existing endpoints and enhance them to be able to prompt - possibly adding new endpoints to enhance this.

This obviously doesn't work for Docker APIv2. My proposal was force Docker APIv2 to only use `docker.io` but @vrothberg had objections to that.

**Comment by mheon** [comment]
I'll label this as feature but I'm not really sure, it could also be construed as a bug.

**Comment by rhatdan** [comment]
We have a short term bug, and long term feature.

**Comment by matejvasek** [comment]
I tried this with older podman v3.2.3 and the bug is not present. Was it introduced just lately?

**Comment by mheon** [comment]
It's possible we swapped the default from "permissive" to "enforcing" recently?

**Comment by matejvasek** [comment]
@mheon actually it's possible that I changed it recently manually on my machine :man_facepalming:

**Comment by rhatdan** [comment]
Fedora 34 defaults to enforcing as well.  If you fully remove containers-common and reinstall it on f34 you will get enforcing mode.

**Comment by vrothberg** [comment]
> This obviously doesn't work for Docker APIv2. My proposal was force Docker APIv2 to only use `docker.io` but @vrothberg had objections to that.

I am more supportive of this idea now.  The only remark I have is that the project-atomic docker did use search registries for pulling but since that can be a security issue, I am feel more inclined to just enforce docker.io.

@mtrmac what do you think?

**Comment by mtrmac** [comment]
TL;DR: Is this bug about keeping old clients _working_ or about making them _secure_? It seems that it was about keeping them _working_, in which case hard-coding `docker.io` seems counterproductive; they can already do that via configuration.

---

Longer version:

I’m, also, mildly, supportive of shipping such configuration, i.e. at least the _upstream_ Podman configuration, with `unqualified-search-registries = ["docker.io"]`, if it is not doing that already. (Fedora can override with their own legacy behavior if they want, but with shipped short-names, that could go away as well.) That might well be very controversial, especially if it included Fedora/RHEL, and probably undesirable on upgrades if the goal is to keep clients working.

It might also make sense to for the API server to have a “docker.io only” boolean, enabled by default… Both this boolean, or the hard-coded behavior, assumes users who don’t want search via the API, _and_ want the search via the CLI. It’s very plausible that there are many users who want either of these behaviors _individually_, but it feels so weird to me to think they would want both, so inconsistent philosophies, _together_.

My default approach is not to break existing clients, but we don’t know whether the client is a `projectatomic/docker`-behavior client or an upstream-behavior client (do we?).

Overall I don’t have a strong opinion.

---

Details, probably not necessary to read:

I _assume_ that there are basically three classes of users:
- Upstream users of Docker, with upstream configuration. There, the short names in the API are always `docker.io`
- Fedora/RHEL users of `projectatomic/docker`, with the shipped multi-registry configuration. There, the short names have always involved search; and clients that did convert `docker.io/$short` names to `$short` in the API had already had to be patched to work against such servers, for many years.
- Oblivious clients that trigger search unintentionally, because clients convert inputs to API requests via `FamiliarString`.

[That could well be wrong; are users routinely mixing upstream code with Fedora configurations, or the opposite? Something else?]

The upstream users, and `projectatomic/docker` users with upstream-like behavior expectations, should configure Podman to search only `docker.io`, (and keep the enforcing policy) and AFAICS the issue goes away entirely.

The `projectatomic/docker` that have built API clients around the search behavior users could be severely broken by changing the behavior, in principle.
- (Most importantly, they should change their client’s image names, or the clients, to send fully-qualified names. If they do that, whatever we do with the Podman server doesn’t matter.)
- For those that want the search, or don’t want to change anything in clients, AFAICS hard-coding `docker.io` is _worse_ for compatibility than having a _configurable_ `enforcing` behavior (which they can turn back to `permissive`). If it motivates them to change the clients, it’s a net win for security; if it motivates them not to upgrade to Podman, it’s probably a loss.

WRT the oblivious users, who do submit fully-qualified names, or don’t expect the search behavior at all because they were designed against upstream behavior, and are completely unaware of clients using short names, and that turning to search. I guess all hinges on how large _this_ group is. If it were an overwhelming majority, the API server should default to `docker.io` only, or at least allow that as an option (possibly a _new_ option for users that want the search on the CLI but don’t realize the API triggers search??). I have no data at all on how large a problem this is.

Arguably it’s confusing to document “if you are a Docker CE user, never enable permissive mode, but configure search registries; if you are a search-behavior-dependent, enable permissive mode”; but I don’t think taking away the options is going to confuse the two classes of users, taken together, much less. The third group is unreachable by documentation, so it might make sense to prioritize their expectations a bit [OTOH it’s somewhat a bit of bad precedent to punish careful understanding by breaking behavior for those who carefully read the documentation to keep code working for users who don’t take that effort].

Changing the shipped upstream default to search only `docker.io` would, at least, remove one set of confused users because the defaults would just work; _and_ it would help the oblivious; OTOH it would generate more work for CLI users that rely on search+enforcing currently. Assuming it’s easier to change CLI habits than to change API clients (is it??), removing the default search behavior by hard-coding `docker.io` again is somewhat attractive because it simplifies the documentation _and_ over time migrates users away from relying on the problematic feature entirely.

**Comment by matejvasek** [comment]
> The decision was to (continue) enforcing docker.io as the only search registry in `podman machine init`

@vrothberg I created machine yesterday and VM's `/etc/containers/registries.conf` contains
```
unqualified-search-registries = ["registry.fedoraproject.org", "registry.access.redhat.com", "docker.io", "quay.io"]
```
Shouldn't it be just [ "docker.io" ] ?

**Comment by vrothberg** [comment]
> What is release cycle of the image for podman machine? Right now it has `v3.3.1`, when it will contain `v3.4.2`? I am asking because of some other feature that probably will be in `v3.4.2`.

Testing stream should have 3.4: https://twitter.com/tnk4on/status/1450941030572781569
